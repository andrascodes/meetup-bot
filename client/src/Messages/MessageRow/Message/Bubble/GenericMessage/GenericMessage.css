.GenericPagingButton {
  background-color: Transparent;
  background-repeat:no-repeat;
  border: none;
  cursor:pointer;
  overflow: hidden;
  outline:none;
}

.GenericPagingButton:hover,
.GenericPagingButton:active,
.GenericPagingButton:focus {
  color: #fff;
  outline: none;
}