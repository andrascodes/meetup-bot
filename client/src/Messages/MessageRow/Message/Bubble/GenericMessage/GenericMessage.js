import React, { Component } from 'react'

import { GenericElement } from './GenericElement'

import './GenericMessage.css'

export class GenericMessage extends Component {
  state = {
    index: 0,
    elements: this.props.attachment.payload.elements
  }

  incrementIndex = e => {
    e.persist()
    e.preventDefault()
    if(this.state.index < this.state.elements.length - 1) {
      this.setState(state => ({
        index: state.index + 1
      }))
    }
  }

  decrementIndex = e => {
    e.persist()
    e.preventDefault()
    if(this.state.index > 0) {
      this.setState(state => ({
        index: state.index - 1
      }))
    }
  }

  render() {
    return (
      <div className="GenericMessage row">
        <div className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-left">
          <button 
            className={`GenericPagingButton btn ${this.state.index <= 0 ? 'disabled' : ''}`}
            onClick={this.decrementIndex}
          >
            <i className="fa fa-chevron-left" />
          </button>
        </div>
        
        <div className="col-lg-8 col-md-8 col-sm-8 col-xs-8 text-center">
          <GenericElement 
            index={this.state.index} 
            element={this.state.elements[this.state.index]} 
          />
        </div>
        
        <div className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-right">
          <button 
            className={`GenericPagingButton btn ${this.state.index === this.state.elements.length - 1 ? 'disabled' : ''}`} 
            onClick={this.incrementIndex} 
          >
            <i className="fa fa-chevron-right" />
          </button>
        </div>
      </div>
    )
  }
}