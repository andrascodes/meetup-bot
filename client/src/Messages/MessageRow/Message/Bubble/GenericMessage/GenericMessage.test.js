import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import { GenericMessage } from './GenericMessage'

describe(`GenericMessage`, () => {

  const elements = [
    {
      image_url: 'https://www.google.com',
      title: 'Title1',
      subtitle: 'Subtitle1',
      default_action: {
        url: 'https://www.google.com'
      },
      buttons: [
        {
          title: 'Btn1 title1',
        },
        {
          title: 'Btn1 title2',
        },
      ]
    },
    {
      image_url: 'https://www.google.com',
      title: 'Title2',
      subtitle: 'Subtitle2',
      default_action: {
        url: 'https://www.google.com'
      },
      buttons: [
        {
          title: 'Btn2 title1',
        },
        {
          title: 'Btn2 title2',
        },
      ]
    }
  ]
  
  // attachement: payload: elements
  test('renders first element from elements by default', () => {
    const wrapper = mount(
      <GenericMessage 
        attachment={{
          payload: {
            elements,
          }
        }}
      />
    )
    expect(wrapper.find('.GenericElementTitle').text()).toEqual(elements[0].title)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('renders second element when clicked on next button', () => {
    const wrapper = mount(
        <GenericMessage 
          attachment={{
            payload: {
              elements,
            }
          }}
        />
      )
      wrapper.find('.GenericPagingButton > .fa-chevron-right').parent().simulate('click', { 
        persist: () => {}, 
        preventDefault: () => {} 
      })
      expect(wrapper.find('.GenericElementTitle').text()).toEqual(elements[1].title)
      expect(toJson(wrapper)).toMatchSnapshot()
  })
  
  test('renders first element when clicked on previous button', () => {
    const wrapper = mount(
      <GenericMessage 
        attachment={{
          payload: {
            elements,
          }
        }}
      />
    )
    wrapper.find('.GenericPagingButton > .fa-chevron-right').parent().simulate('click', { 
      persist: () => {}, 
      preventDefault: () => {} 
    })
    wrapper.find('.GenericPagingButton > .fa-chevron-left').parent().simulate('click', { 
      persist: () => {}, 
      preventDefault: () => {} 
    })
    expect(wrapper.find('.GenericElementTitle').text()).toEqual(elements[0].title)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

})