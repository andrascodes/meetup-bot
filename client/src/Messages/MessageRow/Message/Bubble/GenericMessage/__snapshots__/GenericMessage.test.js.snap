exports[`GenericMessage renders first element from elements by default 1`] = `
<GenericMessage
  attachment={
    Object {
      "payload": Object {
        "elements": Array [
          Object {
            "buttons": Array [
              Object {
                "title": "Btn1 title1",
              },
              Object {
                "title": "Btn1 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle1",
            "title": "Title1",
          },
          Object {
            "buttons": Array [
              Object {
                "title": "Btn2 title1",
              },
              Object {
                "title": "Btn2 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle2",
            "title": "Title2",
          },
        ],
      },
    }
  }>
  <div
    className="GenericMessage row">
    <div
      className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-left">
      <button
        className="GenericPagingButton btn disabled"
        onClick={[Function]}>
        <i
          className="fa fa-chevron-left" />
      </button>
    </div>
    <div
      className="col-lg-8 col-md-8 col-sm-8 col-xs-8 text-center">
      <Component
        element={
          Object {
            "buttons": Array [
              Object {
                "title": "Btn1 title1",
              },
              Object {
                "title": "Btn1 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle1",
            "title": "Title1",
          }
        }
        index={0}>
        <div
          className="GenericElement">
          <img
            alt=""
            className="GenericElementImage img-responsive"
            src="https://www.google.com" />
          <div
            className="GenericElementTitleContainer GenericDataContainer text-left">
            <div
              className="GenericElementTitle">
              Title1
            </div>
            <div
              className="GenericElementSubtitle">
              Subtitle1
            </div>
            <div
              className="GenericElementDefaultAction GenericElementSubtitle">
              www.google.com
            </div>
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn1 title1
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn1 title2
          </div>
        </div>
      </Component>
    </div>
    <div
      className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-right">
      <button
        className="GenericPagingButton btn "
        onClick={[Function]}>
        <i
          className="fa fa-chevron-right" />
      </button>
    </div>
  </div>
</GenericMessage>
`;

exports[`GenericMessage renders first element when clicked on previous button 1`] = `
<GenericMessage
  attachment={
    Object {
      "payload": Object {
        "elements": Array [
          Object {
            "buttons": Array [
              Object {
                "title": "Btn1 title1",
              },
              Object {
                "title": "Btn1 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle1",
            "title": "Title1",
          },
          Object {
            "buttons": Array [
              Object {
                "title": "Btn2 title1",
              },
              Object {
                "title": "Btn2 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle2",
            "title": "Title2",
          },
        ],
      },
    }
  }>
  <div
    className="GenericMessage row">
    <div
      className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-left">
      <button
        className="GenericPagingButton btn disabled"
        onClick={[Function]}>
        <i
          className="fa fa-chevron-left" />
      </button>
    </div>
    <div
      className="col-lg-8 col-md-8 col-sm-8 col-xs-8 text-center">
      <Component
        element={
          Object {
            "buttons": Array [
              Object {
                "title": "Btn1 title1",
              },
              Object {
                "title": "Btn1 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle1",
            "title": "Title1",
          }
        }
        index={0}>
        <div
          className="GenericElement">
          <img
            alt=""
            className="GenericElementImage img-responsive"
            src="https://www.google.com" />
          <div
            className="GenericElementTitleContainer GenericDataContainer text-left">
            <div
              className="GenericElementTitle">
              Title1
            </div>
            <div
              className="GenericElementSubtitle">
              Subtitle1
            </div>
            <div
              className="GenericElementDefaultAction GenericElementSubtitle">
              www.google.com
            </div>
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn1 title1
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn1 title2
          </div>
        </div>
      </Component>
    </div>
    <div
      className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-right">
      <button
        className="GenericPagingButton btn "
        onClick={[Function]}>
        <i
          className="fa fa-chevron-right" />
      </button>
    </div>
  </div>
</GenericMessage>
`;

exports[`GenericMessage renders second element when clicked on next button 1`] = `
<GenericMessage
  attachment={
    Object {
      "payload": Object {
        "elements": Array [
          Object {
            "buttons": Array [
              Object {
                "title": "Btn1 title1",
              },
              Object {
                "title": "Btn1 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle1",
            "title": "Title1",
          },
          Object {
            "buttons": Array [
              Object {
                "title": "Btn2 title1",
              },
              Object {
                "title": "Btn2 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle2",
            "title": "Title2",
          },
        ],
      },
    }
  }>
  <div
    className="GenericMessage row">
    <div
      className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-left">
      <button
        className="GenericPagingButton btn "
        onClick={[Function]}>
        <i
          className="fa fa-chevron-left" />
      </button>
    </div>
    <div
      className="col-lg-8 col-md-8 col-sm-8 col-xs-8 text-center">
      <Component
        element={
          Object {
            "buttons": Array [
              Object {
                "title": "Btn2 title1",
              },
              Object {
                "title": "Btn2 title2",
              },
            ],
            "default_action": Object {
              "url": "https://www.google.com",
            },
            "image_url": "https://www.google.com",
            "subtitle": "Subtitle2",
            "title": "Title2",
          }
        }
        index={1}>
        <div
          className="GenericElement">
          <img
            alt=""
            className="GenericElementImage img-responsive"
            src="https://www.google.com" />
          <div
            className="GenericElementTitleContainer GenericDataContainer text-left">
            <div
              className="GenericElementTitle">
              Title2
            </div>
            <div
              className="GenericElementSubtitle">
              Subtitle2
            </div>
            <div
              className="GenericElementDefaultAction GenericElementSubtitle">
              www.google.com
            </div>
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn2 title1
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn2 title2
          </div>
        </div>
      </Component>
    </div>
    <div
      className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-right">
      <button
        className="GenericPagingButton btn disabled"
        onClick={[Function]}>
        <i
          className="fa fa-chevron-right" />
      </button>
    </div>
  </div>
</GenericMessage>
`;
